src/main/java/net/dv8tion/jda/internal/entities/MessageReference.java

```diff
@@ -0,0 +1,117 @@
+package net.dv8tion.jda.internal.entities;
+
+import net.dv8tion.jda.api.JDA;
+import net.dv8tion.jda.api.entities.Message;
+import net.dv8tion.jda.api.entities.MessageType;
+import net.dv8tion.jda.api.entities.TextChannel;
+import net.dv8tion.jda.api.requests.RestAction;
+import net.dv8tion.jda.internal.JDAImpl;
+import net.dv8tion.jda.internal.requests.RestActionImpl;
+import net.dv8tion.jda.internal.requests.Route;
+import org.jetbrains.annotations.Nullable;
+
+import javax.annotation.Nonnull;
+
+public class MessageReference
+{
+    private final long messageId;
+    private final long channelId;
+    private final Message referencedMessage;
+    private final JDA api;
+
+    public MessageReference(long messageId, long channelId, @Nullable Message referencedMessage, JDA api)
+    {
+        this.messageId = messageId;
+        this.channelId = channelId;
+        this.referencedMessage = referencedMessage;
+        this.api = api;
+    }
+
+    /**
+     * Retrieves the referenced message for this message. If one is already present, it will be returned.
+     *
+     * @throws java.lang.IllegalStateException
+     *         If this message does not have a reference
+     *
+     * @return {@link net.dv8tion.jda.api.requests.RestAction RestAction} - Type: {@link net.dv8tion.jda.api.entities.Message}
+     */
+    @Nonnull
+    public RestAction<Message> retrieveReferencedMessage() {
```

Comment:
```suggestion
    public RestAction<Message> retrieveReferencedMessage()
    {
```

Reply:
How about just `resolve()` instead of this long name? `message.getMessageReference().retrieveReferencedMessage()` seems way too long and unnecessarily redundant.